compat/socket.cc

```diff
@@ -0,0 +1,166 @@
+        { WSAEISCONN, EISCONN },
+
+        // values checked by ignoreErrno()
+        { WSAEWOULDBLOCK, EWOULDBLOCK },
+        // WSAEAGAIN not defined
+        { WSAEALREADY, EALREADY },
+        { WSAEINTR, EINTR },
+        // WSARESTART not defined
+
+        // values checked by limitError()
+        { WSAEMFILE, EMFILE },
+        // WSAENFILE not defined
+
+        // values checked by TunnelStateData::Connection::debugLevelForError()
+        { WSAECONNRESET, ECONNRESET }
+    };
+
+    const auto wsa = WSAGetLastError();
+    const auto itr = CodeMap->find(wsa);
+    if (itr != CodeMap->cend())
+        errno = itr->second;
+    else
+        errno = wsa;
+}
+
+int
+xaccept(int socketFd, struct sockaddr *a, socklen_t *l)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (handle == INVALID_HANDLE) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::accept(handle, a, l);
+    if (result == INVALID_SOCKET) {
+        SetErrnoFromWsaError();
+        return SOCKET_ERROR;
+    } else {
+        return _open_osfhandle(result, 0);
+    }
+}
+
+int
+xbind(int socketFd, const struct sockaddr * n, socklen_t l)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (handle == INVALID_HANDLE) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::bind(handle,n,l);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xclose(int fd)
+{
+    auto sock = _get_osfhandle(fd);
+    if (sock == INVALID_HANDLE) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+
+    char l_so_type[sizeof(int)];
+    int l_so_type_siz = sizeof(l_so_type);
+    if (::getsockopt(sock, SOL_SOCKET, SO_TYPE, l_so_type, &l_so_type_siz) == 0) {
+        const auto result = closesocket(sock);
+        if (result == SOCKET_ERROR)
+            SetErrnoFromWsaError();
+        return result;
+    } else {
+        const auto result = _close(fd);
+        if (result)
+            SetErrnoFromWsaError();
+        return result;
+    }
+}
+
+int
+xconnect(int socketFd, const struct sockaddr * n, socklen_t l)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (handle == INVALID_HANDLE) {
+        // errno is already set by _get_osfhandle()
+        return SOCKET_ERROR;
+    }
+    const auto result = ::connect(handle,n,l);
+    if (result == SOCKET_ERROR)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+struct hostent *
+xgethostbyname(const char *name)
+{
+    auto result = ::gethostbyname(name);
+    if (!result)
+        SetErrnoFromWsaError();
+    return result;
+}
+
+int
+xsetsockopt(int socketFd, int l, int o, const void *v, socklen_t n)
+{
+    const auto handle = _get_osfhandle(socketFd);
+    if (handle == INVALID_HANDLE) {
+        // errno is already set by _get_osfhandle()
+        return EBADF;
```

Comment:
This is wrong return result. `setsockopt(2)` API returns `-1` (or `SOCKET_ERROR`) same as all the other functions in here on any errors.

Reply:
corrected. thanks